File: Python/Recursion/permute_string.py

def permute_string(s):
    """Finds all permutations of an input string. Each character is treated as unique
    whether or not the input string contains duplicate characters."""
    chars_list = list(s)
    output = []

    # find all permutations by recursion
    def permute(_list, idx, n, permutations):
        if idx > n - 2:
            permutations += [''.join(_list)]
        else:
            for i in range(idx, len(_list)):
                new_char_list = swap(_list, idx, i)
                permute(new_char_list, idx + 1, n, permutations)

    # swap two characters
    def swap(_list, i, j):
        # copy _list
        li = _list.copy()
        # swap
        temp = li[i]
        li[i] = li[j]
        li[j] = temp

        return li

    permute(chars_list, 0, len(s), output)
    return output

File: JavaScript/Strings/permute_string.js
/**
 * Finds all permutations of a string. Each character is treated as unique
 * whether or not the input string contains duplicate characters. This permutation
 * is without repetition.
 * @param {*} s input string
 * @returns an array of all permutations of the string
 */
export default function permuteString(s) {
  // array of string characters
  let charArr = s.split('');
  // array of all permutations
  let permutations = [];

  // find permutations by recursion
  function permute(arr, i, n) {
    if (i > n - 2) {
      permutations.push(arr.join(''));
    } else {
      for (let j = i; j < n; j++) {
        let swappedCharArr = swap(arr, i, j);
        permute(swappedCharArr, i + 1, n);
      }
    }
  }

  // swap two characters
  function swap(arr, i, j) {
    // make a copy of arr to prevent mutation
    let a = [...arr];
    let temp = a[i];
    a[i] = a[j];
    a[j] = temp;
    return a;
  }

  permute(charArr, 0, s.length);
  return permutations;
}
